File: content/snippets/scala3/interview_lcci/02/08_linked_list_cycle.scala
object Lcci0208LinkedListCycle:
  final case class ListNode(var value: Int, var next: ListNode | Null = null)

  def buildList(nums: Vector[Int]): ListNode | Null =
    val dummy = ListNode(0, null)
    var tail: ListNode = dummy
    nums.foreach { n =>
      tail.next = ListNode(n, null)
      tail = tail.next.asInstanceOf[ListNode]
    }
    dummy.next

  def hasCycle(head: ListNode | Null): Boolean =
    var slow = head
    var fast = head
    while (fast != null && fast.asInstanceOf[ListNode].next != null) {
      slow = slow.asInstanceOf[ListNode].next
      fast = fast.asInstanceOf[ListNode].next.asInstanceOf[ListNode].next
      if (slow == fast) return true
    }
    false

  @main def main(): Unit =
    val a = buildList(Vector(1, 2, 3, 4))
    var tail = a
    while (tail != null && tail.asInstanceOf[ListNode].next != null) {
      tail = tail.asInstanceOf[ListNode].next
    }
    if (tail != null && a != null && a.asInstanceOf[ListNode].next != null) {
      tail.asInstanceOf[ListNode].next = a.asInstanceOf[ListNode].next
    }

    val b = buildList(Vector(1, 2, 3, 4))

    println(hasCycle(a))
    println(hasCycle(b))

File: content/snippets/go/interview_lcci/02/08_linked_list_cycle.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func buildList(nums []int) *ListNode {
	dummy := &ListNode{}
	tail := dummy
	for _, n := range nums {
		tail.Next = &ListNode{Val: n}
		tail = tail.Next
	}
	return dummy.Next
}

func hasCycle(head *ListNode) bool {
	slow, fast := head, head
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
		if slow == fast {
			return true
		}
	}
	return false
}

func main() {
	a := buildList([]int{1, 2, 3, 4})
	tail := a
	for tail.Next != nil {
		tail = tail.Next
	}
	tail.Next = a.Next

	b := buildList([]int{1, 2, 3, 4})

	fmt.Println(hasCycle(a))
	fmt.Println(hasCycle(b))
}
